File: bundler-packages/resolutions/src/utils/safeFao.js
import { try as trySync, tryPromise } from 'effect/Effect'

/**
 * @typedef {ReturnType<typeof safeFao>} SafeFao
 */

export class ReadFileError extends Error {
	/**
	 * @type {'ReadFileError'}
	 */
	_tag = 'ReadFileError'
	/**
	 * @type {'ReadFileError'}
	 * @override
	 */
	name = 'ReadFileError'
	/**
	 * @param {Error} cause
	 */
	constructor(cause) {
		super(`Read file error: ${cause.message}`, { cause })
	}
}

export class ExistsError extends Error {
	/**
	 * @type {'ExistsError'}
	 */
	_tag = 'ExistsError'
	/**
	 * @type {'ExistsError'}
	 * @override
	 */
	name = 'ExistsError'
	/**
	 * @param {Error} cause
	 */
	constructor(cause) {
		super(`Unable to determine existence: ${cause.message}`, { cause })
	}
}

/**
 * Turns a file access object into a safe file access object that returns effects
 * @param {import("../types.js").FileAccessObject} fao
 * @example
 * ```ts
 * const fao = {
 *   	readFile: readFile,
 * 		readFileSync: readFileSync,
 *		existsSync: existsSync,
 * }
 * const safeFao = safeFao(fao)
 * safeFao.readFileSync('path/to/file', 'utf8') .pipe(
 *   tap((fileContent) => console.log(fileContent)),
 * )
 * ```
 */
export const safeFao = (fao) => {
	return {
		/**
		 * @param {string} path
		 * @param {BufferEncoding} encoding
		 */
		readFile: (path, encoding) => {
			return tryPromise({
				try: () => {
					return fao.readFile(path, encoding)
				},
				catch: (e) => {
					return new ReadFileError(/** @type Error */ (e))
				},
			})
		},
		/**
		 * @param {string} path
		 * @param {BufferEncoding} encoding
		 */
		readFileSync: (path, encoding) => {
			return trySync({
				try: () => {
					return fao.readFileSync(path, encoding)
				},
				catch: (e) => {
					return new ReadFileError(/** @type Error */ (e))
				},
			})
		},
		/**
		 * @param {string} path
		 */
		exists: (path) => {
			return tryPromise({
				try: () => {
					return fao.exists(path)
				},
				catch: (e) => {
					return new ExistsError(/** @type Error */ (e))
				},
			})
		},
		/**
		 * @param {string} path
		 */
		existsSync: (path) => {
			return trySync({
				try: () => {
					return fao.existsSync(path)
				},
				catch: (e) => {
					return new ExistsError(/** @type Error */ (e))
				},
			})
		},
	}
}

File: resolutions-go/utils/safe_fao.go
package utils

import (
	"github.com/williamcory/tevm/go-claude/resolutions-go"
)

// SafeFao wraps a FileAccessObject to provide error handling
type SafeFao struct {
	Fao resolutions.FileAccessObject
}

// NewSafeFao creates a new SafeFao
func NewSafeFao(fao resolutions.FileAccessObject) *SafeFao {
	return &SafeFao{
		Fao: fao,
	}
}

// ReadFile safely reads a file
func (s *SafeFao) ReadFile(path string) (string, error) {
	content, err := s.Fao.ReadFile(path)
	if err != nil {
		return "", resolutions.ReadFileError{Cause: err}
	}
	return content, nil
}

// ReadFileSync safely reads a file synchronously
func (s *SafeFao) ReadFileSync(path string) (string, error) {
	content, err := s.Fao.ReadFileSync(path)
	if err != nil {
		return "", resolutions.ReadFileError{Cause: err}
	}
	return content, nil
}

// Exists checks if a file exists
func (s *SafeFao) Exists(path string) (bool, error) {
	exists, err := s.Fao.Exists(path)
	if err != nil {
		return false, resolutions.ExistsError{Cause: err}
	}
	return exists, nil
}

// ExistsSync checks if a file exists synchronously
func (s *SafeFao) ExistsSync(path string) (bool, error) {
	exists, err := s.Fao.ExistsSync(path)
	if err != nil {
		return false, resolutions.ExistsError{Cause: err}
	}
	return exists, nil
}